File: platforms/paperclip/packages/plugin-enigmagent/src/vault-client.ts
/**
 * EnigmAgent — HTTP client for the local vault REST API.
 *
 * Used by both the secrets provider (server-side) and the plugin worker.
 * Requires Node.js ≥ 18 (global fetch).
 *
 * The vault server must be started separately:
 *   enigmagent-mcp --mode rest --port 3737 --vault ~/.enigmagent/vault.json
 *
 * All requests go to 127.0.0.1 only — the vault is never on the network.
 */

export interface VaultClientConfig {
  host?:      string;   // default: 127.0.0.1
  port?:      number;   // default: 3737
  timeoutMs?: number;   // default: 5000
}

export interface VaultStatus  { status: 'ok' | 'error'; unlocked: boolean }
export interface VaultEntry   { id: string; name: string; domain: string | null; created: string }

// ── Error types ────────────────────────────────────────────────────────────────

export type VaultErrorCode =
  | 'vault_locked'
  | 'not_found'
  | 'no_domain_binding'
  | 'domain_mismatch'
  | 'resolve_error'
  | 'server_unreachable'
  | 'timeout';

export class VaultError extends Error {
  constructor(
    public readonly code: VaultErrorCode,
    message: string,
    public readonly placeholder?: string,
  ) {
    super(message);
    this.name = 'VaultError';
  }
}

// ── Client ─────────────────────────────────────────────────────────────────────

export class VaultClient {
  private readonly baseUrl:   string;
  private readonly timeoutMs: number;

  constructor(cfg: VaultClientConfig = {}) {
    this.baseUrl   = `http://${cfg.host ?? '127.0.0.1'}:${cfg.port ?? 3737}`;
    this.timeoutMs = cfg.timeoutMs ?? 5_000;
  }

  // ── Internal fetch wrapper ─────────────────────────────────────────────────

  private async request<T>(method: 'GET' | 'POST', path: string, body?: unknown): Promise<T> {
    const ctrl  = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    try {
      const res = await fetch(`${this.baseUrl}${path}`, {
        method,
        signal: ctrl.signal,
        headers: body ? { 'Content-Type': 'application/json' } : undefined,
        body:    body ? JSON.stringify(body) : undefined,
      });

      const data = await res.json() as T | { error: string; message?: string };

      if (!res.ok) {
        const err = data as { error: string; message?: string };
        throw new VaultError(
          (err.error as VaultErrorCode) ?? 'resolve_error',
          err.message ?? `HTTP ${res.status}`,
        );
      }
      return data as T;
    } catch (err: unknown) {
      if (err instanceof VaultError) throw err;
      const isAbort = err instanceof Error && err.name === 'AbortError';
      throw new VaultError(
        isAbort ? 'timeout' : 'server_unreachable',
        isAbort
          ? `EnigmAgent vault timed out after ${this.timeoutMs}ms`
          : `EnigmAgent vault unreachable at ${this.baseUrl} — run: ` +
            `enigmagent-mcp --mode rest --port ${this.baseUrl.split(':')[2]} ` +
            `--vault ~/.enigmagent/vault.json`,
      );
    } finally {
      clearTimeout(timer);
    }
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  /** Check whether the vault server is running and unlocked. */
  async getStatus(): Promise<VaultStatus> {
    return this.request<VaultStatus>('GET', '/status');
  }

  /** List all secrets — names and domains only, never values. */
  async listSecrets(): Promise<VaultEntry[]> {
    const data = await this.request<{ entries: VaultEntry[] }>('GET', '/list');
    return data.entries;
  }

  /**
   * Resolve a single placeholder to its real value.
   *
   * @param placeholder  Name without braces: `GITHUB_TOKEN`, `LOGIN:github.com`, `DOC:policy.md`
   * @param origin       Requesting origin for domain-binding checks. Use `http://localhost`
   *                     for server-side resolution within Paperclip.
   */
  async resolve(placeholder: string, origin: string): Promise<string> {
    const data = await this.request<{ value: string }>('POST', '/resolve', { placeholder, origin });
    if (typeof data.value !== 'string') {
      throw new VaultError('resolve_error', 'Server returned no value', placeholder);
    }
    return data.value;
  }

  /**
   * Resolve multiple placeholders in parallel, capturing per-item errors.
   * Returns a Map of placeholder → value | VaultError.
   */
  async resolveBatch(
    items: Array<{ placeholder: string; origin: string }>,
  ): Promise<Map<string, string | VaultError>> {
    const results = new Map<string, string | VaultError>();
    await Promise.all(
      items.map(async ({ placeholder, origin }) => {
        try {
          results.set(placeholder, await this.resolve(placeholder, origin));
        } catch (err) {
          results.set(
            placeholder,
            err instanceof VaultError
              ? err
              : new VaultError('resolve_error', String(err), placeholder),
          );
        }
      }),
    );
    return results;
  }
}

File: platforms/hermes-agent/plugin/enigmagent/vault_client.py
"""
EnigmAgent Hermes Plugin — vault HTTP client.

Communicates with the EnigmAgent REST API running on localhost.
The server must be started separately before Hermes is launched:

    enigmagent-mcp --mode rest --port 3737 --vault ~/.enigmagent/vault.json

Uses only Python standard-library modules (no extra dependencies).
All requests go to 127.0.0.1 — the vault API is never on the network.
"""

from __future__ import annotations

import json
import logging
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Optional

logger = logging.getLogger(__name__)

# ── Error types ────────────────────────────────────────────────────────────────

VAULT_ERROR_CODES = {
    "vault_locked",
    "not_found",
    "no_domain_binding",
    "domain_mismatch",
    "resolve_error",
    "server_unreachable",
    "timeout",
}


class VaultError(Exception):
    """Raised when the EnigmAgent vault returns an error or is unreachable."""

    def __init__(
        self,
        code: str,
        message: str,
        placeholder: Optional[str] = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.placeholder = placeholder

    def __repr__(self) -> str:
        return f"VaultError(code={self.code!r}, placeholder={self.placeholder!r})"


# ── VaultClient ────────────────────────────────────────────────────────────────


class VaultClient:
    """
    Minimal HTTP client for the EnigmAgent local REST API.

    Args:
        host:       API host (default: 127.0.0.1).
        port:       API port (default: 3737).
        timeout_s:  Per-request timeout in seconds (default: 5).
    """

    def __init__(
        self,
        host: str = "127.0.0.1",
        port: int = 3737,
        timeout_s: float = 5.0,
    ) -> None:
        self._base = f"http://{host}:{port}"
        self._timeout = timeout_s

    # ── Low-level request helper ──────────────────────────────────────────────

    def _get(self, path: str) -> dict:
        url = f"{self._base}{path}"
        try:
            with urllib.request.urlopen(url, timeout=self._timeout) as resp:
                data = json.loads(resp.read().decode())
            return data
        except urllib.error.HTTPError as exc:
            body = {}
            try:
                body = json.loads(exc.read().decode())
            except Exception:
                pass
            code = body.get("error", "resolve_error")
            msg  = body.get("message", str(exc))
            raise VaultError(code, msg) from exc
        except (urllib.error.URLError, OSError) as exc:
            raise VaultError(
                "server_unreachable",
                f"EnigmAgent server unreachable at {self._base} — is it running? "
                f"Start with: enigmagent-mcp --mode rest --port {self._base.split(':')[-1]} "
                f"--vault ~/.enigmagent/vault.json",
            ) from exc
        except TimeoutError as exc:
            raise VaultError(
                "timeout",
                f"EnigmAgent server timed out after {self._timeout}s",
            ) from exc

    def _post(self, path: str, payload: dict) -> dict:
        url  = f"{self._base}{path}"
        data = json.dumps(payload).encode()
        req  = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as exc:
            body = {}
            try:
                body = json.loads(exc.read().decode())
            except Exception:
                pass
            code = body.get("error", "resolve_error")
            msg  = body.get("message", str(exc))
            raise VaultError(code, msg) from exc
        except (urllib.error.URLError, OSError) as exc:
            raise VaultError(
                "server_unreachable",
                f"EnigmAgent server unreachable at {self._base}",
            ) from exc
        except TimeoutError as exc:
            raise VaultError("timeout", f"Timed out after {self._timeout}s") from exc

    # ── Public API ────────────────────────────────────────────────────────────

    def get_status(self) -> dict:
        """
        Check whether the vault server is running and the vault is unlocked.

        Returns:
            {"status": "ok", "unlocked": bool}

        Raises:
            VaultError with code="server_unreachable" if the server is not running.
        """
        return self._get("/status")

    def list_secrets(self) -> list[dict]:
        """
        List all secrets by name and domain — never returns actual values.

        Returns:
            List of {"id": str, "name": str, "domain": str|None, "created": str}

        Raises:
            VaultError with code="vault_locked" if vault is locked.
        """
        data = self._get("/list")
        return data.get("entries", [])

    def resolve(self, placeholder: str, origin: str) -> str:
        """
        Resolve a ``{{PLACEHOLDER}}`` name to its real value.

        Args:
            placeholder:  The name without braces, e.g. ``GITHUB_TOKEN``,
                          ``LOGIN:github.com``, ``DOC:report.md``.
            origin:       The requesting origin for domain-binding checks.
                          Use ``http://localhost`` for local Hermes tool calls.

        Returns:
            The decrypted secret value as a plain string.

        Raises:
            VaultError with appropriate code on failure.
        """
        data = self._post("/resolve", {"placeholder": placeholder, "origin": origin})
        value = data.get("value")
        if value is None:
            raise VaultError("resolve_error", "Server returned no value")
        return value

    def resolve_batch(
        self,
        items: list[tuple[str, str]],
        max_workers: int = 8,
    ) -> dict[str, str | VaultError]:
        """
        Resolve multiple ``(placeholder, origin)`` pairs in parallel.

        Returns:
            Mapping of placeholder → resolved_value or VaultError.
            Errors are captured per-placeholder — a single failure
            does not block the rest.
        """
        results: dict[str, str | VaultError] = {}

        with ThreadPoolExecutor(max_workers=min(max_workers, len(items))) as pool:
            futures = {
                pool.submit(self.resolve, placeholder, origin): placeholder
                for placeholder, origin in items
            }
            for fut in as_completed(futures):
                placeholder = futures[fut]
                try:
                    results[placeholder] = fut.result()
                except VaultError as exc:
                    results[placeholder] = exc
                except Exception as exc:
                    results[placeholder] = VaultError("resolve_error", str(exc), placeholder)

        return results
